presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,537 @@
+import com.facebook.presto.orc.stream.LongInputStream;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockLease;
+import com.facebook.presto.spi.block.ClosingBlockLease;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import com.facebook.presto.spi.block.VariableWidthBlock;
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.airlift.units.DataSize;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.array.Arrays.ensureCapacity;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DATA;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.LENGTH;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.PRESENT;
+import static com.facebook.presto.orc.reader.SliceSelectiveReader.computeTruncatedLength;
+import static com.facebook.presto.orc.reader.SliceSelectiveReader.getMaxCodePointCount;
+import static com.facebook.presto.orc.stream.MissingInputStreamSource.missingStreamSource;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SliceDirectSelectiveStreamReader
+        implements SelectiveStreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SliceDirectSelectiveStreamReader.class).instanceSize();
+    private static final int ONE_GIGABYTE = toIntExact(new DataSize(1, GIGABYTE).toBytes());
+
+    private final TupleDomainFilter filter;
+    private final boolean nonDeterministicFilter;
+    private final boolean nullsAllowed;
+
+    private final StreamDescriptor streamDescriptor;
+    private final Type outputType;
+
+    private int readOffset;
+
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    private BooleanInputStream presentStream;
+    private InputStreamSource<ByteArrayInputStream> dataStreamSource = missingStreamSource(ByteArrayInputStream.class);
+    private ByteArrayInputStream dataStream;
+    private InputStreamSource<LongInputStream> lengthStreamSource = missingStreamSource(LongInputStream.class);
+    private LongInputStream lengthStream;
+
+    private boolean rowGroupOpen;
+    private LocalMemoryContext systemMemoryContext;
+
+    private boolean[] nulls;
+
+    protected int[] outputPositions;
+    protected int outputPositionCount;
+
+    protected final boolean outputRequired;
```

Comment:
- outputRequired and allNulls should be private
- why initialize allNulls to true? I think default values of false is more appropriate

Reply:
made them private but lost the change while reverting some local changes. yes allNulls was always false. was testing a local change and got left